File: js/sdk-react-native/ios/ArmenusModelView.swift
import Foundation
import SceneKit
import UIKit

/**
 Inline 3D preview on iOS.

 SceneKit, not RealityKit. RealityKit is the newer engine and the wrong tool
 for a card in a scrolling list: it spins up an ARView with a render loop and a
 session per instance, which is heavy for a preview that is not doing AR, and
 it has no cheap "just show me this model on a plain background" path. SCNView
 is a plain UIView that draws a scene and stops when it is off screen.

 USDZ only. SceneKit has no glTF importer, which is why `EmbedModel.usdzUrl`
 gates the preview on iOS and not only the AR button.
 */
@objc(ArmenusModelView)
public final class ArmenusModelView: UIView {

  private let sceneView = SCNView()
  private var modelNode: SCNNode?
  private var currentSource: String?

  @objc public var onModelLoad: (() -> Void)?
  @objc public var onModelError: ((String) -> Void)?

  /* -- props -------------------------------------------------------------- */

  @objc public var source: NSString = "" {
    didSet { loadIfNeeded() }
  }

  @objc public var cameraOrbitTheta: Double = 0 { didSet { applyCamera() } }
  @objc public var cameraOrbitPhi: Double = 75 { didSet { applyCamera() } }
  @objc public var cameraDistance: Double = 1.05 { didSet { applyCamera() } }

  @objc public var exposure: Double = 1 {
    didSet { sceneView.scene?.rootNode.light?.intensity = CGFloat(exposure * 1000) }
  }

  @objc public var shadowIntensity: Double = 1 { didSet { applyShadow() } }

  @objc public var autoRotate: Bool = true { didSet { applyAutoRotate() } }

  @objc public var interactionEnabled: Bool = true {
    didSet { sceneView.allowsCameraControl = interactionEnabled }
  }

  /* -- lifecycle ---------------------------------------------------------- */

  public override init(frame: CGRect) {
    super.init(frame: frame)
    setUp()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setUp()
  }

  private func setUp() {
    sceneView.frame = bounds
    sceneView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    sceneView.backgroundColor = .clear
    sceneView.antialiasingMode = .multisampling4X
    // Physically-based lighting, so a dish lit for PBR does not come out flat.
    sceneView.autoenablesDefaultLighting = true
    sceneView.allowsCameraControl = interactionEnabled

    /*
     Paused until a model is loaded, and paused again when the view leaves the
     window. Each SCNView owns a CADisplayLink; a feed with twenty dish cards
     that all keep rendering off screen will drain a battery and drop the
     scroll below 60fps for no visible benefit.
     */
    sceneView.rendersContinuously = false
    sceneView.isPlaying = false

    addSubview(sceneView)
  }

  public override func didMoveToWindow() {
    super.didMoveToWindow()
    sceneView.isPlaying = window != nil && autoRotate && modelNode != nil
  }

  /* -- loading ------------------------------------------------------------ */

  private func loadIfNeeded() {
    let path = source as String
    guard !path.isEmpty, path != currentSource else { return }
    currentSource = path

    // A local path from `prefetch` loads synchronously off disk; a remote URL
    // is routed through the same cache so the two paths cannot diverge.
    if path.hasPrefix("http") {
      ArmenusModelCache.shared.file(for: path) { [weak self] local, error in
        guard let self else { return }
        DispatchQueue.main.async {
          if let local {
            self.loadScene(from: local)
          } else {
            self.onModelError?(error?.localizedDescription ?? "Download failed")
          }
        }
      }
    } else {
      loadScene(from: URL(fileURLWithPath: path))
    }
  }

  private func loadScene(from url: URL) {
    /*
     Parsed off the main thread. A textured dish is a few MB of USDZ and
     decoding it inline drops frames in whatever list the card is sitting in —
     the single most visible performance mistake available here.
     */
    DispatchQueue.global(qos: .userInitiated).async { [weak self] in
      do {
        let scene = try SCNScene(url: url, options: [.checkConsistency: false])
        DispatchQueue.main.async {
          guard let self else { return }
          self.install(scene: scene)
        }
      } catch {
        DispatchQueue.main.async {
          self?.onModelError?(error.localizedDescription)
        }
      }
    }
  }

  private func install(scene: SCNScene) {
    sceneView.scene = scene

    let node = SCNNode()
    for child in scene.rootNode.childNodes {
      node.addChildNode(child)
    }
    scene.rootNode.addChildNode(node)
    modelNode = node

    /*
     Re-centre on the bounding box.

     Exporters disagree about origin as wildly as they do about scale — a dish
     can arrive with its pivot metres away from the mesh. Without this the
     camera frames empty space and the preview looks broken rather than
     mis-framed.
     */
    let (minBound, maxBound) = node.boundingBox
    let centre = SCNVector3(
      (minBound.x + maxBound.x) / 2,
      (minBound.y + maxBound.y) / 2,
      (minBound.z + maxBound.z) / 2)
    node.pivot = SCNMatrix4MakeTranslation(centre.x, centre.y, centre.z)

    applyCamera()
    applyShadow()
    applyAutoRotate()

    sceneView.isPlaying = window != nil && autoRotate
    onModelLoad?()
  }

  /* -- framing ------------------------------------------------------------ */

  private func applyCamera() {
    guard let node = modelNode else { return }

    let (minBound, maxBound) = node.boundingBox
    let extent = SCNVector3(
      maxBound.x - minBound.x, maxBound.y - minBound.y, maxBound.z - minBound.z)
    let radius = max(extent.x, max(extent.y, extent.z)) / 2
    let distance = Float(cameraDistance) * radius * 2.6

    // Spherical to Cartesian, matching model-viewer's convention so that one
    // `cameraOrbit` string frames the dish identically on web and native.
    let theta = Float(cameraOrbitTheta) * .pi / 180
    let phi = Float(cameraOrbitPhi) * .pi / 180

    let camera = sceneView.pointOfView ?? {
      let node = SCNNode()
      node.camera = SCNCamera()
      sceneView.scene?.rootNode.addChildNode(node)
      sceneView.pointOfView = node
      return node
    }()

    camera.position = SCNVector3(
      distance * sin(phi) * sin(theta),
      distance * cos(phi),
      distance * sin(phi) * cos(theta))
    camera.look(at: SCNVector3Zero)
    camera.camera?.zNear = Double(max(0.01, distance * 0.01))
    camera.camera?.zFar = Double(distance * 10)
  }

  private func applyShadow() {
    guard let scene = sceneView.scene else { return }

    let existing = scene.rootNode.childNode(withName: "armenus-key", recursively: false)
    let node = existing ?? SCNNode()
    node.name = "armenus-key"

    let light = node.light ?? SCNLight()
    light.type = .directional
    light.castsShadow = shadowIntensity > 0
    light.shadowMode = .deferred
    light.shadowRadius = 8
    light.shadowColor = UIColor.black.withAlphaComponent(CGFloat(0.4 * shadowIntensity))
    node.light = light
    node.eulerAngles = SCNVector3(-Float.pi / 3, Float.pi / 5, 0)

    if existing == nil { scene.rootNode.addChildNode(node) }
  }

  private func applyAutoRotate() {
    guard let node = modelNode else { return }
    node.removeAction(forKey: "armenus-spin")

    guard autoRotate else {
      sceneView.isPlaying = false
      return
    }

    let spin = SCNAction.repeatForever(
      SCNAction.rotateBy(x: 0, y: .pi * 2, z: 0, duration: 18))
    node.runAction(spin, forKey: "armenus-spin")
    sceneView.isPlaying = window != nil
  }
}

File: js/sdk-react-native/android/src/main/java/app/armenus/sdk/ArmenusModelView.kt
package app.armenus.sdk

import android.content.Context
import android.view.MotionEvent
import android.view.Choreographer
import android.widget.FrameLayout
import io.github.sceneview.SceneView
import io.github.sceneview.node.ModelNode
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.math.cos
import kotlin.math.sin

/**
 * Inline 3D preview on Android.
 *
 * Filament, through SceneView. GLB only — Filament has no USDZ importer, which
 * is the mirror of SceneKit's missing glTF one and the reason every model in
 * this system is stored in both formats.
 *
 * No ARCore session is created. This is a plain view that draws a mesh; AR is
 * a separate handoff to Scene Viewer.
 */
class ArmenusModelView(context: Context) : FrameLayout(context) {

  private val sceneView = SceneView(context)
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
  private var modelNode: ModelNode? = null
  private var currentSource: String? = null
  private var loadJob: Job? = null
  private var disposed = false

  var onModelLoad: (() -> Unit)? = null
  var onModelError: ((String) -> Unit)? = null

  /* -- props -------------------------------------------------------------- */

  var cameraOrbitTheta: Double = 0.0
    set(value) { field = value; applyCamera() }

  var cameraOrbitPhi: Double = 75.0
    set(value) { field = value; applyCamera() }

  var cameraDistance: Double = 1.05
    set(value) { field = value; applyCamera() }

  var autoRotate: Boolean = true
    set(value) {
      field = value
      if (value) startSpin() else stopSpin()
    }

  var interactionEnabled: Boolean = true

  init {
    addView(sceneView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
  }

  fun setSource(source: String) {
    if (source.isEmpty() || source == currentSource) return
    currentSource = source

    loadJob?.cancel()
    stopSpin()
    modelNode?.let { sceneView.removeChildNode(it); it.destroy() }
    modelNode = null
    loadJob = scope.launch {
      try {
        /*
         * Fetch and parse off the main thread. A textured dish is a few MB and
         * Filament's glTF parse is not cheap — doing it inline drops frames in
         * whatever list the card is scrolling in, which is the most visible
         * performance mistake available on this path.
         */
        val buffer = withContext(Dispatchers.IO) {
          val file = if (source.startsWith("http")) {
            ArmenusModelCache.file(context, source)
          } else {
            java.io.File(source)
          }
          ArmenusGlb.prepare(file.readBytes())
        }

        if (disposed || currentSource != source) return@launch
        val instance = sceneView.modelLoader.createModelInstance(
          java.nio.ByteBuffer.wrap(buffer),
        )
        val node = ModelNode(
          modelInstance = instance,
          autoAnimate = false,
          scaleToUnits = 1.0f,
          centerOrigin = io.github.sceneview.math.Position(0f, 0f, 0f),
        )

        sceneView.addChildNode(node)
        modelNode = node
        applyCamera()
        if (autoRotate) startSpin()
        onModelLoad?.invoke()
      } catch (error: CancellationException) {
        throw error
      } catch (error: Throwable) {
        onModelError?.invoke(error.message ?: "Model failed to load")
      }
    }
  }

  /* -- framing ------------------------------------------------------------ */

  private fun applyCamera() {
    // Spherical to Cartesian, matching model-viewer's convention so that one
    // `cameraOrbit` string frames a dish identically on web, iOS and Android.
    val theta = Math.toRadians(cameraOrbitTheta)
    val phi = Math.toRadians(cameraOrbitPhi)
    val distance = cameraDistance * 2.6

    sceneView.cameraNode.position = io.github.sceneview.math.Position(
      x = (distance * sin(phi) * sin(theta)).toFloat(),
      y = (distance * cos(phi)).toFloat(),
      z = (distance * sin(phi) * cos(theta)).toFloat(),
    )
    sceneView.cameraNode.lookAt(io.github.sceneview.math.Position(0f, 0f, 0f))
  }

  /* -- spin --------------------------------------------------------------- */

  private var spinCallback: Choreographer.FrameCallback? = null

  private fun startSpin() {
    stopSpin()
    val node = modelNode ?: return
    var last = 0L

    /*
     * Driven by Choreographer rather than a fixed timer, so the rotation is
     * tied to actual frame delivery. A timer keeps firing while the view is
     * off screen or the device is throttling, which burns battery and makes
     * the spin jump when it comes back.
     */
    val callback = object : Choreographer.FrameCallback {
      override fun doFrame(frameTimeNanos: Long) {
        if (last != 0L) {
          val delta = (frameTimeNanos - last) / 1_000_000_000f
          // One revolution every 18 seconds, matching the web and iOS viewers.
          node.rotation = node.rotation.copy(
            y = node.rotation.y + delta * (360f / 18f),
          )
        }
        last = frameTimeNanos
        if (spinCallback === this) Choreographer.getInstance().postFrameCallback(this)
      }
    }
    spinCallback = callback
    Choreographer.getInstance().postFrameCallback(callback)
  }

  private fun stopSpin() {
    spinCallback?.let { Choreographer.getInstance().removeFrameCallback(it) }
    spinCallback = null
  }

  override fun onInterceptTouchEvent(event: MotionEvent): Boolean = !interactionEnabled

  override fun onTouchEvent(event: MotionEvent): Boolean = false

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    if (!disposed && autoRotate) startSpin()
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    // React Native can temporarily detach a view while recycling a list.
    stopSpin()
  }

  fun dispose() {
    if (disposed) return
    disposed = true
    stopSpin()
    scope.cancel()
    modelNode?.let { sceneView.removeChildNode(it); it.destroy() }
    modelNode = null
    sceneView.destroy()
  }
}
